### PR TITLE
ionic-build 1.0.0

### PR DESCRIPTION
**Pull request by patrickbussmann**
### New Pull Request Checklist

*Please mark the points which you did / accept.*

- [X] I read the [Step Development Guideline](https://github.com/bitrise-io/bitrise/blob/master/_docs/step-development-guideline.md)
- [X] I have a test for my Step, which can be run with `bitrise run test` (in the step's repository)
- [X] I did run `bitrise run audit-this-step` (in the step's repository - note, if you don't have this workflow in your `bitrise.yml`, [you can copy it from the step template](https://github.com/bitrise-steplib/step-template/blob/master/bitrise.yml).)
- [X] I read and accept the [Abandoned Step policy](https://github.com/bitrise-io/bitrise-steplib#abandoned-step-policy)

Yay first bitrise step finished within 24 hours :+1: 
I hope everything is fine and I help the community with this.

https://ionicframework.com/
Ionic is a big framework for mobile development and I think it is really useful.
I created a documentation how to build the Ionic conference app from scratch with Bitrise.io :smile: 
You can find it here: https://github.com/bussmann-it/steps-ionic/blob/master/README.md
I referenced it in the step description.

So all people can build Ionic apps quick :smiley: 

This is a example of bitrise.yml for Ionic conference app.
https://github.com/bussmann-it/ionic-conference-app/blob/master/bitrise.yml

With this the users can build their apps really easy with Bitrise.io. Adding the repository. Copy the bitrise.yml. Change 2-5 settings and let's wait for use the built app.

### CODE REVIEW

**Comment by viktorbenei** [comment]
Wow, thanks here again @patrickbussmann for such an awesome, detailed PR! :)

Question 1: are you sure that the constant here https://github.com/bussmann-it/steps-ionic/blob/master/step.sh#L52 is always the same? (`1D6058900D05DD3D006BFB54`)

Question 2:

> With this the users can build their apps really easy with Bitrise.io. Adding the repository. Copy the bitrise.yml. Change 2-5 settings and let's wait for use the built app.

This could be even shorter, if we'd have a scanner for it! With that the project type could be detected as Ionic right away, and the user would get the right `bitrise.yml` automatically after the registration (just like with the scanners we have right now for (native) iOS, Android, macOS, Xamarin, Cordova, .. projects)

We have this on our "short term" task list, to have Ionic project support in the scanners. But for that we'll need a "verified" step (as a policy the scanners can only use "verified" steps).

@patrickbussmann would you be open to move this step into https://github.com/bitrise-community ? We would of course add you to this organization as well, so we can collaborate on steps easier there. You can find more info about this community github org at https://discuss.bitrise.io/t/bitrise-steps-community/496

File: steps/ionic-build/1.0.0/step.yml [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+published_at: 2017-05-14T02:06:53.343409014+02:00
+source:
+  git: https://github.com/bussmann-it/steps-ionic.git
+  commit: d587cee45caaf377f5fb29c0c1590475f2f1a056
+host_os_tags:
+- osx-10.9
+- osx-10.10
+- osx-10.11
+- osx-10.12
+project_type_tags:
```
Comment (viktorbenei) [inline steps/ionic-build/1.0.0/step.yml]: Should match with the guideline: https://github.com/bitrise-io/bitrise/blob/master/_docs/step-development-guideline.md#step-grouping-convention

File: steps/ionic-build/1.0.0/step.yml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+- ios
+- android
+- wp8
+- browser
+type_tags:
```
Comment (viktorbenei) [inline steps/ionic-build/1.0.0/step.yml]: Should match with the guideline: https://github.com/bitrise-io/bitrise/blob/master/_docs/step-development-guideline.md#step-grouping-convention

File: steps/ionic-build/1.0.0/step.yml (hunk 1 above)
Comment (viktorbenei) [inline steps/ionic-build/1.0.0/step.yml]: Well, actually, you could simply remove `project_type_tags` or specify an empty list for it (`[]`). That means "it works with any kind of project type". Until we have a specific project type for ionic, it's easier to let this step show up for any project in the step lists (IMO)

File: steps/ionic-build/1.0.0/step.yml (hunk 2 above)
Comment (viktorbenei) [inline steps/ionic-build/1.0.0/step.yml]: Probably should be `build`

**Comment by viktorbenei** [comment]
@trapacska & @godrei please chime in and review! ;)

**Comment by viktorbenei** [comment]
OFFTOPIC: btw so many `1` and `0` -> `ionic-build 1.0.0 #1000` :P

this is the #1000 PR here!! :tada:

**Comment by patrickbussmann** [comment]
Hello @viktorbenei,

here your answers.

**Answer 1:**
I'm sure that the constant is always the same in XCode 8.*
At least for Ionic projects :smile:
In my Bitrise.io account there are 4 projects (= 8 apps, 4 x iOS, 4x Android)
And all iOS projects using the same constant.
I tested this with the Ionic conference app which is not modified. (Only changed iOS App ID and added bitrise.yml)
And everything works with this.

So yes I'm sure and its tested :+1: 
But maybe it will change in future.
I can test with XCode 7 if you want (and if I find it :laughing:)

**Answer 2:**
I really like your idea with the scanner.
Maybe you remember me. As I started with Bitrise.io I thought I ask you via support chat if there is something for Ionic and you said its planned 🤗
This was in January or February 2017. The chat says 17w ago.

I can put it in my calendar for the next weekend :+1: 
https://github.com/bitrise-io/devcenter/blob/master/docs/tips-and-tricks/write-your-own-scanner.md


And yes, of course, @viktorbenei, we can move this step into https://github.com/bitrise-community.
Will do you do this? Or do you add myself to the organization as you said and I create the repository and everything?


I will start correcting your comments now.

**Comment by viktorbenei** [comment]
> This was in January or February 2017. The chat says 17w ago.

Yeah, it was planned to be released sooner, but something always creeped to the top of our task list.. But Cordova is now out, ReactNative scanner almost finished, and the next one on the list is Ionic! ;)

> And yes, of course, @viktorbenei, we can move this step into https://github.com/bitrise-community.

Awesome! We'll send you the invites tomorrow. What's your bitrise.io username? We'll add you to the community org there too ;)

**Comment by patrickbussmann** [comment]
Yes no problem @viktorbenei ^^
In this case I speeded up your task list so that you can concentrate on other things :+1: 

My bitrise.io username is `patrick.bussmann`.

Thanks :smile: 
If you add me tomorrow I'll transfer the repository. :+1: 


Are my corrections okay now? Commited few hours ago.
Or some changes left in your eyes?

**Comment by viktorbenei** [comment]
> Are my corrections okay now?

LGTM, but @godrei & @trapacska will add the final 👍 / 👎 once they can review it ;)

**Comment by patrickbussmann** [comment]
No problem. It's a :sunny: day and weekend (sunday) so we shouldn't work :laughing: 

Let's continue tomorrow with @godrei and @trapacska :+1:

**Comment by viktorbenei** [comment]
@patrickbussmann just invited you into the GitHub and Bitrise community organizations

Sorry for the delay, yesterday was.. super busy ;)

**Comment by godrei** [comment]
Hi @patrickbussmann,

thanks for the PR!

I would have a questions:

- May i ask you, why do you prefer to use ionic-cli for 'preparing the project' and then use xcode-archive step to build the project?

Doing this way, you build your project two times (for first when you prepare the .xcodeproj by running [ionic build](https://github.com/bussmann-it/steps-ionic/blob/master/step.sh#L27)), then the step xcode-archive builds the project again.
To achive this you also need to modify the generated .xcodeproj file's [code sign settings](https://github.com/bussmann-it/steps-ionic/blob/master/step.sh#L51).

When i was working on [cordova-archive](https://github.com/bitrise-community/steps-cordova-archive) step i decided to build the project by calling `cordova build`, where i pass the `build.json` file to generate the desired signed output artifacts. This way we only need to build the project only once and we do not need to modify the generated project. Another reason is that, this way the scanner can generate much [smaller workflows](https://github.com/bitrise-core/bitrise-init/blob/master/_tests/integration/cordova_test.go#L96) for cordova projects.

I know that this way you do not have the freedom of xcode-archive step (forcing code sign settings, export options, etc...), but do you think it is a real-life need or can you agree that configuring the code signin with `build.json` file is enought?

If you think it would be better to go with your version, then i would recommand to update the step's name, since it just partially builds the project (it build the android part fully, and exports the generated artifacts (.apk files), but just prepares the ios project (by building it using ionic commands)). I would expect to generate outputs from a builder step. 

So i think it would be better if the step would only prepare an ionic project and let our builder steps (xcode-archive, gradle-runner) to build the final artifacts, or do the whole build in the step.

**Comment by patrickbussmann** [comment]
Absolutly no problem, @viktorbenei.
I moved it now and you can find it here: https://github.com/bitrise-community/steps-ionic

I hope everything I did was correct :smile: 

Now I create an update for the steplib and the plugin so that the links are correct for the new location.
Then you can accept PR after review :+1:

**Comment by patrickbussmann** [comment]
@godrei:
Thanks for your review.
At the moment the answer is that it was better for me using the xcode-archive because of the reasons you mentioned. Forcing team, forcing provisioning profile, ...
The signing is the main problem.

The Android build is similar but more complete.
1. Build debug with Ionic/Cordova
2. Sign APK

In iOS I had a lot of troubles with the signing because some things didnt work and some others are working wrong.
And with the current solution I was able to build everything without any problems.
As you can see in the https://github.com/bitrise-community/steps-ionic/blob/master/README.md file - there is a really detailed how to build the sample Ionic conference app with Bitrise.io.
Including generating keystore for signing the APK and getting certificates for signing the IPA. (or build them signed)


I will check if it is possible to add the parameters to ionic build so that we can skip the xcode-build step.
I mean this will be perfect :+1: 
But at the time of developing this step it wasn't so great.

I will check your cordova build because normally Ionic using Cordova for the build process.
But my goal is that everyone can build their apps without problems.


Let me check and I will inform you about results :+1:

**Comment by godrei** [comment]
Hi @patrickbussmann ,

as i mentioned previously we have a cordova-archive step which builds and exports the generated artifacts (ipa, dsym, app, apk) without using any others steps. I was able to create similar builder step for ionic projects: [ionic-archive](https://github.com/bitrise-community/steps-ionic-archive) by only modifying the called commands from `cordova` to `ionic "cordova" "--no-interactive"`.

The new step also works great with the ionic conference app you mentioned in your step's README.

If you agree with that i will close this PR and we will use the [ionic-archive](https://github.com/bitrise-community/steps-ionic-archive) step to build ionic projects, would you like to keep your version in the bitrise-community? 

Thanks for your PR and all of the infos you shared with us, any feedback for the ionic-archive step is well appreciated!

**Comment by patrickbussmann** [comment]
Hmm @godrei, it's okay I think.
If your ionic-archive step works as well as mine with the ability to select cordova and ionic versions then you can close this and ignore it.

I think its important to select between Ionic 2 and Ionic 3 because Ionic 3 was released this/last month so many peoples are using Ionic 2 currently.

I thought my Readme was also better than your default one now. But we can copy mine to ionic-archive if you allow :smile:  https://github.com/bitrise-community/steps-ionic


If @viktorbenei think the same then you can do it as you like. I will pull the ionic-archive and will support them in future if we decide to use them.

**Comment by viktorbenei** [comment]
@patrickbussmann after discussing the matter with @godrei quite thoroughly, I'd support the new `ionic-archive` step for one reason, because that's almost identical to the already existing Cordova archive step. This means lower maintenance overhead, which means more time for other improvements and well as possible shared code between the two steps.

Our plan is to start with that step (ionic archive), and if there would be any issue indicating that's a sub-optimal choice we can still opt to change it to your ionic builder step. From what we know right now this seems to be the best choice long term, but there's a reason why every Bitrise Step is versioned, allowing breaking changes when required ;)

Last but not least I'd like to **thank you** @patrickbussmann for your collaboration, for your previous and future contributions, the discussion with you helped **a lot** in adding Ionic support, which will help everyone using Bitrise for Ionic project! :)

**Comment by patrickbussmann** [comment]
Ok then I will support the ionic-archive step :+1: 

Am I allowed to add the support for selecting Ionic version?
I will change my build flow soon and then I must have the choice between 2 and 3 :smile: 

Should I work with Pull requests then or can I work directly on master or new branch?

**Comment by godrei** [comment]
@patrickbussmann i just started to work on this feature :)
Please let me finish this and i will assign you as reviewer for the pr, what do you think?

**Comment by godrei** [comment]
@patrickbussmann the [master branches are protected](https://discuss.bitrise.io/t/bitrise-steps-community/496/23?u=godreikrisztian) in the bitrise-community org, so you can work with PRs.

I assigned you as a reviewer for the [version selection PR](https://github.com/bitrise-community/steps-ionic-archive/pull/2).

**Comment by viktorbenei** [comment]
To be precise, you can create any branch in the repos if you're in the community gh org, but pushes to master are protected and require a PR+review before merge.

**Pull request closed without merging**